[BugFix] Ranger view/MV authorization for external tables

love-star opened the pull request:
## What I'm doing:

Fixes #72910

## 1. Problem Description

In Ranger (external access controller) mode, `ColumnPrivilege.check()` mishandled `View` and `MaterializedView` objects.

## 2. Code Changes

File modified: `fe/fe-core/src/main/java/com/starrocks/authorization/ColumnPrivilege.java`

### 2.1 Route Views and MVs explicitly in the Ranger branch

Inside `check()`, the `tableUsedExternalAccessController.contains(tableName)` branch now dispatches by table type instead of unconditionally calling `checkColumnAction`:

```java
if (tableUsedExternalAccessController.contains(tableName)) {
    if (table instanceof View view) {
        checkViewPrivilege(context, tableName, view);            // NEW
    } else if (table.isMaterializedView()) {
        Authorizer.checkMaterializedViewAction(...);             // NEW (resource = materialized_view)
    } else {
        // existing column-level check for regular tables
    }
}
```

This ensures Ranger receives the correct resource type (`view` /`materialized_view`) for each object kind.

### 2.2 Extract a public `checkViewPrivilege` helper

The security-view handling logic, previously inlined only in the native branch, is now a single `public static` method shared by both branches:

```java
public static void checkViewPrivilege(ConnectContext context,
                                      TableName tableName, View view) {
    try {
        // connector views are checked as tables
        if (view.isConnectorView()) {
            Authorizer.checkTableAction(context, tableName, PrivilegeType.SELECT);
            return;
        }

        // security views: recursively verify SELECT on all referenced objects
        if (view.isSecurity()) {
            for (TableName t : view.getTableRefs()) {
                BasicTable basicTable = GlobalStateMgr.getCurrentState()
                        .getMetadataMgr()
                        .getBasicTable(context, t.getCatalog(), t.getDb(), t.getTbl());
                if (basicTable.isOlapView()) {
                    QueryStatement qs = ((View) basicTable).getQueryStatement();
                    Analyzer.analyze(qs, context);
                    Authorizer.check(qs, context);
                } else if (basicTable.isMaterializedView()) {
                    Authorizer.checkMaterializedViewAction(context, t, PrivilegeType.SELECT);
                } else {
                    Authorizer.checkTableAction(context, t, PrivilegeType.SELECT);
                }
            }
        }

        // and SELECT on the view itself
        Authorizer.checkViewAction(context, tableName, PrivilegeType.SELECT);
    } catch (AccessDeniedException e) {
        AccessDeniedException.reportAccessDenied(
                tableName.getCatalog(),                          // FIXED: was hard-coded internal catalog
                context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
                PrivilegeType.SELECT.name(), ObjectType.VIEW.name(), tableName.getTbl());
    }
}
```

Both the native branch and the Ranger branch now delegate to this method, so security view semantics are applied uniformly regardless of access controller.

### 2.3 Behavior comparison

| Scenario                          | Before                                       | After                                                       |
|-----------------------------------|----------------------------------------------|-------------------------------------------------------------|
| Ranger + normal View              | `checkColumnAction` (resource = `table`)     | `checkViewAction` (resource = `view`)                       |
| Ranger + Materialized View        | `checkColumnAction` (resource = `table`)     | `checkMaterializedViewAction` (resource = `materialized_view`) |
| Ranger + Security View            | No pass-through to base tables               | Recursive SELECT check on all base tables/views/MVs + the view itself |
| Native + Security View            | Inline implementation                        | Same behavior, now shared via `checkViewPrivilege`          |
| Catalog reported on view denial   | Hard-coded internal catalog                  | Uses `tableName.getCatalog()` (correct for external catalogs) |

## 3. Core Unit Tests

File modified: `fe/fe-core/src/test/java/com/starrocks/authorization/ranger/RangerInterfaceTest.java`

### 3.1 Test fixtures (added in `beforeClass`)

Three objects are pre-created for view/MV/security-view coverage:

- `db.test_view` — regular OLAP view referencing `db.t1`
- `db.test_mv` — materialized view referencing `db.t1`
- `db.test_security_view` — view marked as security via `setSecurity(true)`

### 3.2 Retained tests (5 cases, one per independent code path)

| # | Test method                                                              | Path under test                          | Key assertions                                                                                          |
|---|--------------------------------------------------------------------------|------------------------------------------|---------------------------------------------------------------------------------------------------------|
| 1 | `testColumnPrivilegeCheckViewWithExternalAccessController`               | Ranger + View, allow                     | Ranger plugin is actually invoked with resource type `view` (proves the bypass is fixed)                |
| 2 | `testColumnPrivilegeCheckViewDeniedWithExternalAccessController`         | Ranger + View, deny                      | `ErrorReportException` is thrown on Ranger denial                                                       |
| 3 | `testColumnPrivilegeCheckMaterializedViewWithExternalAccessController`   | Ranger + MV, allow                       | Ranger plugin is invoked with resource type `materialized_view`                                         |
| 4 | `testColumnPrivilegeCheckMaterializedViewDeniedWithExternalAccessController` | Ranger + MV, deny                    | `ErrorReportException` is thrown on Ranger denial                                                       |
| 5 | `testCheckViewPrivilegeSecurityViewWithRanger`                           | Ranger + Security View (core fix)        | Ranger receives BOTH the underlying base table `t1` (resource = `table`) AND the view `test_security_view` (resource = `view`) |

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.1
  - [x] 4.0
  - [x] 3.5

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=56316) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=56316) before we can accept your contribution.<br/><hr/>**haoyuhua** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=56316) it.</sub>

CelerData-Reviewer commented:
@codex review

love-star commented on fe/fe-core/src/main/java/com/starrocks/authorization/ColumnPrivilege.java:
```diff
@@ -131,16 +131,41 @@ public static void check(ConnectContext context, QueryStatement stmt, List<Table
+                        if (table.getType() == Table.TableType.HIVE_VIEW) {
+                            Authorizer.checkTableAction(context, tableName, PrivilegeType.SELECT);
+                        } else {
+                            Authorizer.checkViewAction(context, tableName, PrivilegeType.SELECT);
```
I use table.isConnectorView() to avoid it

love-star commented on fe/fe-core/src/main/java/com/starrocks/authorization/ColumnPrivilege.java:
```diff
@@ -128,22 +128,43 @@ public static void check(ConnectContext context, QueryStatement stmt, List<Table
-
                 if (table instanceof View) {
                     try {
                         // for privilege checking, treat connector view as table
```
keep same with old code: line 207               } else if (table.isMaterializedView()) {

love-star commented:
> Hi team,
> 
> Are there any plans to review/update/merge this PR? Our team is actively experiencing this exact issue in production on StarRocks 3.5.12 using Ranger 2.7.0.
> 
> This bug causes a complete security bypass. In our deployment, any authenticated user—even one with 0 grants and no Ranger policies can `SELECT` from any view across all databases.
> 
> As noted in this PR, `ColumnPrivilege.java` skips `checkViewAction()` entirely when `access_control = ranger`.
> 
> This is a critical blocker for using StarRocks + Ranger in production. We would greatly appreciate it if this fix could be revived and merged into the upcoming releases.

I fix this problem again on the latest branch, if you need this feature, you can cherrypick it to your branch.

eshishki commented:
@HangyuanLiu this is a high-severity security issue (silent authorization bypass for views — and partial misrouting for MVs — when Ranger is enabled). See #72910 root cause walkthrough by @love-star. Author's been waiting on review for 14 months on the original report (#56315) and 3 months on this re-submission.

**Bug:** in `ColumnPrivilege.check()` line 133-145, the `tableUsedExternalAccessController` branch always does column-level Ranger checks regardless of object type. `tableUsedExternalAccessController` triggers per-catalog when an `ExternalAccessController` is configured — so this hits both per-Hive Ranger plugins and global Ranger on default_catalog (`access_control = ranger`), which makes native views and MVs affected too.

For **views**: (a) Ranger gets resource `{"table": "view_name"}` instead of `{"view": "view_name"}` — view policies never match; (b) after optimizer expansion `LogicalScanOperator` holds the base table, not the View, so `tableObjToTableName.get(table)` returns `null`, `scanColumns` ends up with a `null` key, the per-column loop runs over an empty set and **no privilege checks happen at all** for the view's `TableName`. End result: any user can SELECT from any view through Ranger.

For **MVs**: Ranger gets resource `{"table": "mv_name"}` instead of `{"materialized_view": "mv_name"}` — MV-targeted policies never match.

**Fix:** type discrimination in the external-AC branch (mirrors what the native branch already does at line 147-191) — route Views to `checkViewAction` / `checkTableAction`, MVs to `checkMaterializedViewAction`, regular tables to column-level. +333 LOC of tests.

@love-star — CI metadata is failing on three mechanical things:
1. `title-check`: title is 81 chars, limit is usually 70. Shorten — e.g. `[BugFix] Ranger view/MV authorization for external tables`.
2. `backport-check-subtask (3.5/4.0/4.1)` + parent: cherry-pick likely conflicts because `ColumnPrivilege.java` has shifted in those branches. Either uncheck the version boxes where you can't cleanly backport, or prepare manual backport PRs after this lands.
3. `RUN CHECKER`: should go green automatically once the above are fixed.

love-star commented:
@eshishki Thanks a lot for your suggestions! I have shortened the PR title accordingly and passed the title check successfully. However, even after I removed the old backport related labels, the backport check workflow still keeps running automatically every time.

eshishki commented:
> @eshishki However, even after I removed the old backport related labels, the backport check workflow still keeps running automatically every time.

check first box - I have checked the version labels which the pr will be auto-backported to the target branch
uncheck - 4.1

love-star commented:
@eshishki Thanks for your help! I have completed the following operations as required:
1. Checked the checkbox - "I have checked the version labels which the pr will be auto-backported to the target branch";
2. Unchecked the 4.1 version label for auto-backporting.
However, the CI pipeline still throws the error "You must mark this checkbox".  I suspect the root cause is that this PR was created quite early, and the existing backport-check pipeline logic (designed based on older rules) is no longer compatible with the current code state.

love-star commented:
@HangyuanLiu this PR requires two approvals. If you have a minute, could you help introduce another committer to review it? Thank you so much!

CelerData-Reviewer commented:
@codex review

love-star committed:
[BugFix] Ranger view/MV authorization for external tables

Signed-off-by: love-star <haoyuhua007@gmail.com>

CelerData-Reviewer commented:
@codex review

CelerData-Reviewer commented:
@codex review

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/authorization/ColumnPrivilege.java:
```diff
@@ -204,6 +184,48 @@ public static void check(ConnectContext context, QueryStatement stmt, List<Table
+                    } else if (basicTable.isMaterializedView()) {
+                        Authorizer.checkMaterializedViewAction(context, t, PrivilegeType.SELECT);
+                    } else {
+                        Authorizer.checkTableAction(context, t, PrivilegeType.SELECT);
```
Inconsistent for direct ref-table and indirect ref-table with a intermediate view to check column privilege.

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/authorization/ColumnPrivilege.java:
```diff
@@ -204,6 +184,48 @@ public static void check(ConnectContext context, QueryStatement stmt, List<Table
+        try {
+            // for privilege checking, treat connector view as table
+            if (view.isConnectorView()) {
+                Authorizer.checkTableAction(context, tableName, PrivilegeType.SELECT);
```
It's better to add some comment: for consistency for both `starrocks service type` and `hive service type` ranger privilege check. because hive views are treated as table in ranger with `hive service type`.

love-star replied:
Appreciate your feedback! I'll adjust the code based on your comments and talk to you again later.

jaogoy reviewed (commented):
@love-star Thanks for picking this back up and controbuting the PR — the view bypass is a genuine authorization hole, and routing views/MVs to object-level checks (instead of the column loop) is the right fix. A few things worth confirming / hardening before merge.

### 1. (Design discussion) Should a security view check its base *tables* at column level?

In `checkViewPrivilege`, a security view's refs are dispatched by type:

| ref type | path | recurses? | granularity under Ranger |
| --- | --- | --- | --- |
| internal OLAP view | `Authorizer.check(qs)` | yes | base tables end up **column-level** |
| MV | `checkMaterializedViewAction` | no | object-level |
| table / external table / connector view | `checkTableAction` | no | **table-level only** |

So the granularity drifts with reference depth: the *same* table is column-checked when queried directly, but only table-checked when it is a direct ref of a security view. Under Ranger this effectively **launders column-level policies** — a user without column access to `t1.c1` but with table access to `t1` can read `t1.c1` through the security view.

Since the purpose of the pass-through is to stop users from bypassing base-table permissions, I'd argue it should be consistent: a `table` ref should also be column-checked. One clean option — for a security view, run `Authorizer.check(view.getQueryStatement())` for *all* refs (the path you already take for OLAP-view refs) instead of dispatching per ref type; `ColumnPrivilege` then handles every base table uniformly. It is a behavior tightening, so it hinges on the intended security-view semantics. What's your take?

### 2. Other items

| # | severity | item |
| --- | --- | --- |
| 1 | should-fix | The single try/catch in `checkViewPrivilege` reports every `AccessDeniedException` as `ObjectType.VIEW` + the view name. When a security view's base **table/MV** is denied, the message says "access denied to VIEW &lt;view&gt;" even though the user is really missing privilege on the base object — misleading for troubleshooting. Report TABLE / MATERIALIZED_VIEW for base-object failures, and VIEW only for the view's own `checkViewAction`. |
| 2 | test | The motivating case is Hive connector views/MVs, but the added tests all use internal-catalog objects with Ranger on `default_catalog`, so the `isConnectorView()` branch is never exercised by a real connector view. An integration test there would close the loop. |
| 3 | minor | The fix correctly stops views/MVs from going through `scanColumns`, but the underlying behavior — a view's base table is expanded away, so `tableObjToTableName.get(scanTable)` returns null and the columns land under a `null` key — still happens silently. A short assert/WARN in `ScanColumnCollector.visitLogicalTableScan` on `tableName == null` (plus a one-line comment that views/MVs intentionally don't rely on `scanColumns`) would keep a future caller from hitting the same silent drop. |

Thanks again for the fix!

love-star committed:
[BugFix] Fix view privilege bypass: route security view refs through column-level checks and report base failures with correct ObjectType

Signed-off-by: love-star <haoyuhua007@gmail.com>

love-star commented:
@jaogoy 

Thanks for the thorough review! Here's what I changed in response to each item.

## 1. Security view consistency: column-level checks for all base refs (Design Discussion)

Adopted the suggestion. In `ColumnPrivilege.checkViewPrivilege`, all base refs of a security view are now uniformly dispatched through `Authorizer.check(view.getQueryStatement())` instead of being dispatched per ref type. This way OLAP tables, MVs, and connector views — every base object — go through `ColumnPrivilege` for column-level checks, keeping behavior consistent between the StarRocks and Hive service types (Hive treats views as `TABLE` in Ranger). Added an inline comment explaining the consistency intent so future readers don't reintroduce the per-type dispatch.

## 2. AccessDenied no longer mis-reports object type (should-fix)

Split the single `try/catch` in `checkViewPrivilege` so failures are reported against the right object:

| failure source | reported object type |
| --- | --- |
| view's own `checkViewAction` | `ObjectType.VIEW` |
| base table | `ObjectType.TABLE` (or `ObjectType.COLUMN` on the column-level path) |
| base materialized view | `ObjectType.MATERIALIZED_VIEW` |

No more `"access denied to VIEW <view>"` when the user is actually missing a privilege on a base object.

## 3. Connector-view branch test coverage (test)

Used JMockit `MockUp<View>` to force `isConnectorView()` to return `true`, so the `isConnectorView()` branch is exercised without standing up a real Hive/Iceberg/Paimon catalog. The test also asserts the failure surfaces as `ObjectType.TABLE`, matching the Hive Ranger service-type semantics.

## 4. Silent-drop guard in `ScanColumnCollector` (minor)

In `ScanColumnCollector.visitLogicalTableScan`, when `tableName == null` (the table was expanded away from a view/MV and was never recorded in `tableObjToTableName`), it now skips explicitly. Added a one-line comment noting that views/MVs intentionally don't rely on `scanColumns` and are covered by `checkViewPrivilege` / `checkMaterializedViewAction`, so a future caller won't hit the same silent drop.

---

## Test additions (`RangerInterfaceTest.java`)

- **View / MV under an external Ranger access controller** — both allow and deny paths, asserting Ranger is actually invoked at object granularity.
- **Regular OLAP table** — column-level allow and deny paths, asserting Ranger receives a `column` resource.
- **Security view referencing a base table and a base MV** — Ranger receives the base resource (`table` / `materialized_view`) plus the view resource.
- **Regression for should-fix **: base-table denial must surface as `TABLE` / `COLUMN`; base-MV denial must surface as `MATERIALIZED VIEW`. Neither may be reported as `VIEW`.
- **Regression for test **: with `isConnectorView()` mocked to `true`, denial must surface as `ObjectType.TABLE`.
- **Regression for minor **: with an empty `tableObjToTableName`, `ScanColumnCollector` must skip cleanly — no `null` key, no NPE.
- **Root user** must bypass Ranger even when Ranger denies everything; non-root users still go through Ranger.

thanks again!

love-star committed:
[BugFix] Sync test_security_view expected error to base table ObjectType

Signed-off-by: love-star <haoyuhua007@gmail.com>

jaogoy commented:
@HangyuanLiu Would you like to review it again?

love-star commented:
@jaogoy Hi there, gentle ping on this PR. All review feedback has been addressed and all checks are passing. Could you please advise if there are any remaining blockers for merging this PR?

jaogoy commented:
@love-star You should resolve all the comments above
